### DISCUSSION TITLE
Render UserControl to bmp without a Window

### DISCUSSION THREAD

**Discussion by tkefauver** [AvaloniaUI/Avalonia#16299] [Q&A]
I'm testing rendering UI to images using [this:](https://github.com/AvaloniaUI/Avalonia.HtmlRenderer/issues/10#issuecomment-819492618)
```
void RenderToFile(Control target, string path)
{
    var pixelSize = new PixelSize((int) target.Width, (int) target.Height);
    var size = new Size(target.Width, target.Height);
    using (RenderTargetBitmap bitmap = new RenderTargetBitmap(pixelSize, new Vector(96, 96)))
    {
        target.Measure(size);
        target.Arrange(new Rect(size));
        bitmap.Render(target);
        bitmap.Save(path);
    }
}
```

I need to be able to render arbitrary controls (not attached to windows) and this works fine:
```
var test_view = new Ellipse() { Width = 300, Height = 300, Background = Brushes.Orange };
RenderToFile(test_view,"test.png");
```

Doing the same thing from a xaml-based UserControl like below, does **not** work. It'll just renders an empty black image:
```
var test_view2 = new MyCustomUserControl() { Width = 300, Height = 300, Background = Brushes.Orange };
RenderToFile(test_view2,"test2.png");
```

So my question is, is there anyway to render `test_view2` without attaching it to a window? Like using headless mode or manually initializing it somehow?

**Comment by stevemonaco** [comment]
I think the issue is that you're trying to render `TemplatedControl`s (within the `UserControl`) and you aren't connected to the control trees. This means there's no theme availability and no `ControlTheme`s means no visual rendering. It works for controls like `Ellipse` because they're primitives that are self-rendering.

You can try adding your theme directly to the `target.Styles` collection and see what happens.

**Comment by maxkatz6** [comment]
Even in headless mode you would need to attach it to the window.
You can create an empty Window instance **without showing it** - that will work too.
Or you can temporarily attach it to your window as well, though an empty window would be easier.

If you have TemplatedControl, as @stevemonaco mentioned, then yes - it needs to have styles. Normally they would be automatically attached from window.

**Reply by Upanc** [reply]
still need show()

**Comment by tkefauver** [comment]
@maxkatz6 I should have mentioned its on mobile so I can't make a window that's the problem.

**Reply by stevemonaco** [reply]
Sorry, missed this reply. For anybody else, you don't need to attach to a rooted control tree (ie. TopLevel window or view under App). You simply create and apply the theme. eg.

```cs
yourControl.Styles.Add(new FluentTheme());
```

I assume you can cache and assign the same instance of `FluentTheme` in multiple locations.

**Comment by ds1709** [comment]
There's a trick to make it work. You can create hidden window to render controls inside it. The key points are:
1. You need to call `Show` to load window when it created and then `Hide`. After then you can reuse this window without showing it.
1. After setting your control to content you need to wait it will be loaded before render.
1. After control is loaded (because of host window is hidden) you have to call `Measure` and `Arrange` methods manually.
1. Render control using `RenderTargetBitmap`.

Here's the sample code:
```csharp
public static class ControlRenderer
{
    private static readonly Lazy<ContentControl> renderHost = new(CreateRenderHost);

    public static async Task RenderControlAsync(Control control, Stream targetStream)
    {
        var host = renderHost.Value;

        try
        {
            host.Content = control;

            var tcs = new TaskCompletionSource();

            void OnControlLoaded(object? sender, EventArgs e)
                => tcs.TrySetResult();

            control.Loaded += OnControlLoaded;

            if (control.IsLoaded)
            {
                tcs.TrySetResult();
            }

            await tcs.Task;

            control.Loaded -= OnControlLoaded;

            control.Measure(Size.Infinity);
            control.Arrange(new(control.DesiredSize));

            using var r = new RenderTargetBitmap(new PixelSize((int)control.Bounds.Width, (int)control.Bounds.Height));
            r.Render(control);
            r.Save(targetStream);
        }
        finally
        {
            host.Content = null;
        }
    }

    private static ContentControl CreateRenderHost()
    {
        var window = new Window()
        {
            SystemDecorations = SystemDecorations.None,
            ShowInTaskbar = false,
            ShowActivated = false,
            MinWidth = 0,
            MinHeight = 0,
            MaxWidth = 0,
            MaxHeight = 0,
            Width = 0,
            Height = 0,
            Opacity = 0
        };

        window.Show();
        window.Hide();

        return window;
    }
}
```
